packages/protocol/contracts/layer1/based/TaikoInbox.sol

```diff
@@ -0,0 +1,681 @@
+                TooManyBatches()
+            );
+        }
+
+        BatchParams memory params = abi.decode(_params, (BatchParams));
+
+        {
+            address preconfRouter = resolve(LibStrings.B_PRECONF_ROUTER, true);
+            if (preconfRouter == address(0)) {
+                require(params.proposer == address(0), CustomProposerNotAllowed());
+                params.proposer = msg.sender;
+            } else {
+                require(msg.sender == preconfRouter, NotPreconfRouter());
+                require(params.proposer != address(0), CustomProposerMissing());
+            }
+
+            if (params.coinbase == address(0)) {
+                params.coinbase = params.proposer;
+            }
+        }
+
+        if (params.revertIfNotFirstProposal) {
+            require(state.stats2.lastProposedIn != block.number, NotFirstProposal());
+        }
+
+        // Keep track of last batch's information.
+        Batch storage lastBatch;
+        unchecked {
+            lastBatch = state.batches[(stats2.numBatches - 1) % config.batchRingBufferSize];
+        }
+
+        bool calldataUsed = _txList.length != 0;
+
+        require(calldataUsed || params.numBlobs != 0, BlobNotSpecified());
+
+        (uint64 anchorBlockId, uint64 lastBlockTimestamp) = _validateBatchParams(
+            params,
+            config.maxAnchorHeightOffset,
+            config.maxSignalsToReceive,
+            config.maxBlocksPerBatch,
+            lastBatch
+        );
+
+        // This section constructs the metadata for the proposed batch, which is crucial for
+        // nodes/clients to process the batch. The metadata itself is not stored on-chain;
+        // instead, only its hash is kept.
+        // The metadata must be supplied as calldata prior to proving the batch, enabling the
+        // computation and verification of its integrity through the comparison of the metahash.
+        //
+        // Note that `difficulty` has been removed from the metadata. The client and prover must use
+        // the following approach to calculate a block's difficulty:
+        //  `keccak256(abi.encode("TAIKO_DIFFICULTY", block.number))`
+        meta_ = BatchMetadata({
+            txListHash: calldataUsed
+                ? keccak256(_txList)
+                : _calcTxListHash(params.firstBlobIndex, params.numBlobs),
+            extraData: bytes32(uint256(config.baseFeeConfig.sharingPctg)),
+            coinbase: params.coinbase,
+            batchId: stats2.numBatches,
+            gasLimit: config.blockMaxGasLimit,
+            lastBlockTimestamp: lastBlockTimestamp,
+            parentMetaHash: lastBatch.metaHash,
+            proposer: params.proposer,
+            livenessBond: config.livenessBond,
+            proposedAt: uint64(block.timestamp),
+            proposedIn: uint64(block.number),
+            txListOffset: params.txListOffset,
+            txListSize: params.txListSize,
```

Comment:
Looks like `txListOffset` and `txListSize` are no longer used? I remember that they were used to reuse one single blob in previous design.

Reply:
They should still used by the node and prover. Only `txListSize ` bytes starting at `txListOffset ` in the txList calldata or the concatenated blobs should be used by the node for decoding into a list of txs.